datastore/http/http.go

```diff
@@ -0,0 +1,223 @@
+
+	err := validateConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	transport := http.DefaultTransport
+	if config.TLS != nil {
+		transport, err = getTransport(config.TLS)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	producer := &Producer{
+		namespace:        namespace,
+		metricsCollector: metricsCollector,
+		workerChannels:   make([]chan *telemetry.Record, config.WorkerCount),
+		address:          config.Address,
+		logger:           logger,
+		httpClient: http.Client{
+			Timeout:   time.Duration(config.Timeout) * time.Second,
+			Transport: transport,
+		},
+	}
+
+	for i := 0; i < config.WorkerCount; i++ {
+		producer.workerChannels[i] = make(chan *telemetry.Record, bufferSize)
+		go producer.worker(producer.workerChannels[i])
+	}
+
+	logger.Infof("registered http producer for namespace: %s", namespace)
+	return producer, nil
+}
+
+// validateConfig validates configuration values and sets defaults if value not set
+func validateConfig(c *Config) error {
+	if c.WorkerCount < 0 {
+		return errors.New("invalid http worker count")
+	}
+	if c.WorkerCount == 0 {
+		c.WorkerCount = 5
+	}
+	if c.Timeout < 0 {
+		return errors.New("invalid http timeout")
+	}
+	if c.Timeout == 0 {
+		c.Timeout = 10
+	}
+	return nil
+}
+
+func getTransport(tlsConfig *TLSCertificate) (*http.Transport, error) {
+	clientTLSCert, err := tls.LoadX509KeyPair(tlsConfig.CertFile, tlsConfig.KeyFile)
+	if err != nil {
+		return nil, err
+	}
+	return &http.Transport{
+		TLSClientConfig: &tls.Config{
+			Certificates: []tls.Certificate{clientTLSCert},
+		},
+	}, nil
+}
+
+func (p *Producer) worker(records <-chan *telemetry.Record) {
+	for record := range records {
+		metricsRegistry.bufferSize.Sub(1, map[string]string{})
+		p.sendHTTP(record)
+	}
+}
+
+func (p *Producer) sendHTTP(record *telemetry.Record) {
+	url := fmt.Sprintf("%s?namespace=%s&type=%s", p.address, p.namespace, record.TxType)
+	payload := record.Payload()
+
+	req, err := http.NewRequest("POST", url, bytes.NewBuffer(payload))
+	if err != nil {
+		p.logError(fmt.Errorf("create_request_err %s", err.Error()))
+		return
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+	for key, value := range record.Metadata() {
+		req.Header.Set(key, value)
+	}
+
+	resp, err := p.httpClient.Do(req)
+	if err != nil {
+		p.logError(fmt.Errorf("send_request_err %s", err.Error()))
+		return
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode >= 200 && resp.StatusCode <= 299 {
+		metricsRegistry.produceCount.Inc(map[string]string{"record_type": record.TxType})
+		metricsRegistry.byteTotal.Add(int64(record.Length()), map[string]string{"record_type": record.TxType})
+		return
+	}
+	p.logError(fmt.Errorf("response_status_code %d", resp.StatusCode))
+}
+
+func vinToHash(vin string) uint32 {
+	h := fnv.New32a()
+	h.Write([]byte(vin))
+	return h.Sum32()
+}
+
+func (p *Producer) getRecordChannel(vin string) chan *telemetry.Record {
+	return p.workerChannels[vinToHash(vin)%uint32(len(p.workerChannels))]
```

Comment:
hmm. I agree with https://github.com/teslamotors/fleet-telemetry/pull/91#discussion_r1462620776 but at the same time, it could be possible that not all channels will be used (all vins getting partitioned into a single channel or only one vin to stream and increasing worker channel will not make a difference). I guess this is fine since its not going to be recommended producer but maybe we should add a comment in readme about relation between worker channels and vins associated with this account

Reply:
I agree it's good to call this out. I have added comentary about this to `WorkerCount` in `Config`. We could also add a label to `http_produce_buffer_size` metric calling out size of each worker's buffer. Want me to add this?

Reply:
yes sure!

Reply:
Added :+1: